File: src/main/java/discretemath/structure/set/AbstractSet.java
package discretemath.structure.set;

public abstract class AbstractSet<T> implements Set<T> {

	@Override
	public final int hashCode() {
		if (isInfinite()) {
			RangeSet<?> rangeSet = (RangeSet<?>) this;
			return rangeSet.hashCode();
		} else {
			return iterableHashcode(this);
		}
	}

	@Override
	public final boolean equals(Object o) {
		if (this == o) return true;

		if (!(o instanceof Set otherSet)) {
			return false;
		}

		boolean thisIsInfinite = this.isInfinite();
		boolean otherIsInfinite = otherSet.isInfinite();

		if (thisIsInfinite != otherIsInfinite) { // if one of them is infinite, but second not
			return false;
		}

		if (thisIsInfinite) { // both infinite
			RangeSet<?> thisRangeSet = (RangeSet<?>) this;
			RangeSet<?> otherRangeSet = (RangeSet<?>) this;

			return thisRangeSet.range().equals(otherRangeSet.range());
		} else {
			if (this.size() != otherSet.size()) {
				return false;
			}

			for (Object item : otherSet) {
				@SuppressWarnings("unchecked")
				T casted = (T) item;
				if (!this.contains(casted)) {
					return false;
				}
			}


			return true;
		}
	}

	private static int iterableHashcode(Iterable<?> iterable) {
		int result = 1;

		for (Object element : iterable)
			result = 31 * result + (element == null ? 0 : element.hashCode());

		return result;
	}
}

File: src/main/scala/discretemath/structure/set/main/AbstractSet.scala
package discretemath.structure.set.main

import discretemath.structure.set.RangeSet

trait AbstractSet[T] extends Set[T] {

  final def canEqual(a: Any): Boolean = a.isInstanceOf[Set[_]]

  final override def equals(that: Any): Boolean = {
    that match {
      case otherSet: Set[T] => {
        val thisIsInfinite = this.isInfinite
        val otherIsInfinite = otherSet.isInfinite

        if (thisIsInfinite != otherIsInfinite) { // if one of them is infinite, but second not
          return false
        }

        if (thisIsInfinite) { // both infinite
          val thisRangeSet = this.asInstanceOf[RangeSet[_]]
          val otherRangeSet = this.asInstanceOf[RangeSet[_]]
          return thisRangeSet.range == otherRangeSet.range
        }
        else {
          if (this.cardinality != otherSet.cardinality) {
            return false
          }

          return this.forall(item => otherSet.contains(item))
        }
      }
      case _ => false
    }
  }

  final override def hashCode: Int = AbstractSet.iterableHashcode(this)
}

object AbstractSet {
  private def iterableHashcode(iterable: Iterable[_]) = {
    var result = 1
    for (element <- iterable) {
      result = 31 * result + (if (element == null) 0
      else element.hashCode)
    }
    result
  }
}